libp2p/connectionsHandler.go

```diff
@@ -0,0 +1,389 @@
+		Seeders:                  make([]string, 0),
+		IntraShardValidators:     make(map[uint32][]string),
+		IntraShardObservers:      make(map[uint32][]string),
+		CrossShardValidators:     make(map[uint32][]string),
+		CrossShardObservers:      make(map[uint32][]string),
+		FullHistoryObservers:     make(map[uint32][]string),
+		NumObserversOnShard:      make(map[uint32]int),
+		NumValidatorsOnShard:     make(map[uint32]int),
+		NumPreferredPeersOnShard: make(map[uint32]int),
+	}
+
+	handler.mutPeerResolver.RLock()
+	defer handler.mutPeerResolver.RUnlock()
+
+	selfPeerInfo := handler.peerShardResolver.GetPeerInfo(handler.peerID())
+	connPeerInfo.SelfShardID = selfPeerInfo.ShardID
+
+	for _, p := range peers {
+		conns := handler.p2pHost.Network().ConnsToPeer(p)
+		connString := "[invalid connection string]"
+		if len(conns) > 0 {
+			connString = conns[0].RemoteMultiaddr().String() + "/p2p/" + p.String()
+		}
+
+		pid := core.PeerID(p)
+		peerInfo := handler.peerShardResolver.GetPeerInfo(pid)
+		handler.appendPeerInfo(peerInfo, selfPeerInfo, pid, connPeerInfo, connString)
+
+		if handler.preferredPeersHolder.Contains(pid) {
+			connPeerInfo.NumPreferredPeersOnShard[peerInfo.ShardID]++
+		}
+	}
+
+	return connPeerInfo
+}
+
+func (handler *connectionsHandler) appendPeerInfo(peerInfo, selfPeerInfo core.P2PPeerInfo, pid core.PeerID, connPeerInfo *p2p.ConnectedPeersInfo, connString string) {
+	switch peerInfo.PeerType {
+	case core.UnknownPeer:
+		if handler.sharder.IsSeeder(pid) {
+			connPeerInfo.Seeders = append(connPeerInfo.Seeders, connString)
+		} else {
+			connPeerInfo.UnknownPeers = append(connPeerInfo.UnknownPeers, connString)
+		}
+	case core.ValidatorPeer:
+		connPeerInfo.NumValidatorsOnShard[peerInfo.ShardID]++
+		if selfPeerInfo.ShardID != peerInfo.ShardID {
+			connPeerInfo.CrossShardValidators[peerInfo.ShardID] = append(connPeerInfo.CrossShardValidators[peerInfo.ShardID], connString)
+			connPeerInfo.NumCrossShardValidators++
+		} else {
+			connPeerInfo.IntraShardValidators[peerInfo.ShardID] = append(connPeerInfo.IntraShardValidators[peerInfo.ShardID], connString)
+			connPeerInfo.NumIntraShardValidators++
+		}
+	case core.ObserverPeer:
+		connPeerInfo.NumObserversOnShard[peerInfo.ShardID]++
+		if peerInfo.PeerSubType == core.FullHistoryObserver {
+			connPeerInfo.FullHistoryObservers[peerInfo.ShardID] = append(connPeerInfo.FullHistoryObservers[peerInfo.ShardID], connString)
+			connPeerInfo.NumFullHistoryObservers++
+			break
+		}
+		if selfPeerInfo.ShardID != peerInfo.ShardID {
+			connPeerInfo.CrossShardObservers[peerInfo.ShardID] = append(connPeerInfo.CrossShardObservers[peerInfo.ShardID], connString)
+			connPeerInfo.NumCrossShardObservers++
+			break
+		}
+
+		connPeerInfo.IntraShardObservers[peerInfo.ShardID] = append(connPeerInfo.IntraShardObservers[peerInfo.ShardID], connString)
+		connPeerInfo.NumIntraShardObservers++
+	}
+}
+
+func (handler *connectionsHandler) peerID() core.PeerID {
```

Comment:
same question: can we provide directly the peerID on the connectionsHandler's constructor and avoid this cast & function?